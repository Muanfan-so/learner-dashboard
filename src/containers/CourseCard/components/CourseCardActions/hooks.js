import { handleEvent } from 'data/services/segment/utils';
import { eventNames } from 'data/services/segment/constants';

export const useCourseCardActionData = () => {
  const trackUpgradeClick = () => {
    handleEvent(eventNames.upgradeCourse, {
      pageName: 'learner_home',
      linkType: 'button',
      linkCategory: 'green_upgrade',
    });
  };

  return {
    trackUpgradeClick,
  };
};

export default useCourseCardActionData;
