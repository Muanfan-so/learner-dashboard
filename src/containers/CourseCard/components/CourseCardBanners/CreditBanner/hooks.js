import { StrictDict } from 'utils';
import { hooks as appHooks } from 'data/redux';

import ApprovedContent from './views/ApprovedContent';
import EligibleContent from './views/EligibleContent';
import MustRequestContent from './views/MustRequestContent';
import PendingContent from './views/PendingContent';
import RejectedContent from './views/RejectedContent';

export const statusComponents = StrictDict({
  pending: PendingContent,
  approved: ApprovedContent,
  rejected: RejectedContent,
});

export const useCreditBannerData = (cardId) => {
  const credit = appHooks.useCardCreditData(cardId);
  const { supportEmail } = appHooks.usePlatformSettingsData();
  if (!credit.isEligible) { return null; }

  const { error, purchased, requestStatus } = credit;
  let ContentComponent = EligibleContent;
  if (purchased) {
    if (requestStatus == null) {
      ContentComponent = MustRequestContent;
    } else if (Object.keys(statusComponents).includes(requestStatus)) {
      ContentComponent = statusComponents[requestStatus];
    }
    // Current behavior is to show Elligible State if unknown request status is returned
  }
  return {
    ContentComponent,
    error,
    supportEmail,
  };
};

export default {
  useCreditBannerData,
};
