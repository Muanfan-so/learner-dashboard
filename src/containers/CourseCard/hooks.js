import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

export const useCardData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const { title, bannerImgSrc } = appHooks.useCardCourseData(cardId);
  const { isEnrolled } = appHooks.useCardEnrollmentData(cardId);

  return {
    isEnrolled,
    title,
    bannerImgSrc,
    formatMessage,
  };
};

export default useCardData;
