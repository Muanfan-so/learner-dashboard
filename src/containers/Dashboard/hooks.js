import React from 'react';
import { useWindowSize, breakpoints } from '@edx/paragon';
import { useDispatch } from 'react-redux';
import { thunkActions } from 'data/redux';

export const useIsDashboardCollapsed = () => {
  const { width } = useWindowSize();
  return width < breakpoints.large.maxWidth;
};

export const useInitializeDashboard = () => {
  const dispatch = useDispatch();
  React.useEffect(
    () => { dispatch(thunkActions.app.initialize()); },
    [dispatch],
  );
};

export default {
  useIsDashboardCollapsed,
  useInitializeDashboard,
};
