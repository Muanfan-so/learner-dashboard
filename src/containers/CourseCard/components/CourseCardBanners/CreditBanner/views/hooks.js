import React from 'react';

import { StrictDict } from 'utils';
import { apiHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  creditRequestData: (val) => React.useState(val), // eslint-disable-line
});

export const useCreditRequestData = (cardId) => {
  const [requestData, setRequestData] = module.state.creditRequestData(null);
  const createCreditApiRequest = apiHooks.useCreateCreditRequest(cardId);
  const createCreditRequest = (e) => {
    e.preventDefault();
    createCreditApiRequest().then(setRequestData);
  };
  return { requestData, createCreditRequest };
};

export default {
  useCreditRequestData,
};
