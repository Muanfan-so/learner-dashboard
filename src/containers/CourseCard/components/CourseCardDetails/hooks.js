import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

import * as hooks from './hooks';
import messages from './messages';

export const useAccessMessage = ({ cardId }) => {
  const { formatMessage, formatDate } = useIntl();
  const enrollment = appHooks.useCardEnrollmentData(cardId);
  const courseRun = appHooks.useCardCourseRunData(cardId);
  if (!courseRun.isStarted) {
    if (!courseRun.startDate) { return null; }
    const startDate = formatDate(courseRun.startDate);
    return formatMessage(messages.courseStarts, { startDate });
  }
  if (enrollment.isEnrolled) {
    const { isArchived, endDate } = courseRun;
    const {
      accessExpirationDate,
      isAuditAccessExpired,
    } = enrollment;

    if (enrollment.isAudit && accessExpirationDate) {
      return formatMessage(
        isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
        { accessExpirationDate: formatDate(accessExpirationDate) },
      );
    }
    if (!endDate) { return null; }
    return formatMessage(
      isArchived ? messages.courseEnded : messages.courseEnds,
      { endDate: formatDate(endDate) },
    );
  }
  return null;
};

export const useCardDetailsData = ({ dispatch, cardId }) => {
  const { formatMessage } = useIntl();
  const providerName = appHooks.useCardProviderData(cardId).name;
  const { courseNumber } = appHooks.useCardCourseData(cardId);
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = appHooks.useCardEntitlementData(cardId);

  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: hooks.useAccessMessage({ cardId }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    courseNumber,
    changeOrLeaveSessionMessage: formatMessage(messages.changeOrLeaveSessionButton),
  };
};

export default useCardDetailsData;
