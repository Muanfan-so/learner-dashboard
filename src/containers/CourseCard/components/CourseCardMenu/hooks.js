import React from 'react';
import { StrictDict } from 'utils';

import track from 'tracking';
import { reduxHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val), // eslint-disable-line
  isEmailSettingsVisible: (val) => React.useState(val), // eslint-disable-line
});

export const useUnenrollData = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useEmailSettings = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useHandleToggleDropdown = (cardId) => {
  const eventName = track.course.courseOptionsDropdownClicked;
  const trackCourseEvent = reduxHooks.useTrackCourseEvent(eventName, cardId);
  return (isOpen) => {
    if (isOpen) { trackCourseEvent(); }
  };
};

export const useCourseCardMenu = (cardId) => {
  const { courseName } = reduxHooks.useCardCourseData(cardId);
  const { isEnrolled, isEmailEnabled } = reduxHooks.useCardEnrollmentData(cardId);
  const { twitter, facebook } = reduxHooks.useCardSocialSettingsData(cardId);
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const { isEarned } = reduxHooks.useCardCertificateData(cardId);
  const handleTwitterShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'twitter',
  );
  const handleFacebookShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'facebook',
  );

  const showUnenrollItem = isEnrolled && !isEarned;
  const showDropdown = showUnenrollItem || isEmailEnabled || facebook.isEnabled || twitter.isEnabled;

  return {
    courseName,
    isMasquerading,
    isEmailEnabled,
    showUnenrollItem,
    showDropdown,
    facebook,
    twitter,
    handleTwitterShare,
    handleFacebookShare,
  };
};
