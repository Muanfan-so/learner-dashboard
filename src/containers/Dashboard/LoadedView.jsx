import React from 'react';
import { Container, Col, Row } from '@edx/paragon';

import CourseList from 'containers/CourseList';
import WidgetSidebar from 'containers/WidgetSidebar';
import hooks from './hooks';

export const columnConfig = {
  courseList: {
    sm: { span: 12, offset: 0 },
    md: { span: 10, offset: 1 },
    lg: { span: 12, offset: 0 },
    xl: { span: 8, offset: 0 },
  },
  sidebar: {
    sm: { span: 12, offset: 0 },
    md: { span: 10, offset: 1 },
    lg: { span: 12, offset: 0 },
    xl: { span: 4, offset: 0 },
  },
};

export const LoadedView = () => {
  const isCollapsed = hooks.useIsDashboardCollapsed();

  return (
    <Container fluid size="xl">
      <Row>
        <Col {...columnConfig.courseList} className="course-list-column">
          <CourseList />
        </Col>
        <Col {...columnConfig.sidebar} className="sidebar-column">
          {!isCollapsed && (<h2 className="course-list-title">&nbsp;</h2>)}
          <WidgetSidebar />
        </Col>
      </Row>
    </Container>
  );
};

export default LoadedView;
