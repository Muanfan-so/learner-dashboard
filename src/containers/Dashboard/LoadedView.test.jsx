import { shallow } from 'enzyme';
import { Col, Row } from '@edx/paragon';

import CourseList from 'containers/CourseList';
import WidgetSidebar from 'containers/WidgetSidebar';

import hooks from './hooks';
import LoadedView, { columnConfig } from './LoadedView';

jest.mock('./hooks', () => ({
  useIsDashboardCollapsed: jest.fn(() => true),
}));

describe('LoadedView', () => {
  const testColumns = () => {
    it('loads courseList and sidebar column layout', () => {
      const columns = shallow(<LoadedView />).find(Row).find(Col);
      Object.keys(columnConfig.courseList).forEach(size => {
        expect(columns.at(0).props()[size]).toEqual(columnConfig.courseList[size]);
      });
      Object.keys(columnConfig.sidebar).forEach(size => {
        expect(columns.at(1).props()[size]).toEqual(columnConfig.sidebar[size]);
      });
    });
    it('displays CourseList in first column', () => {
      const columns = shallow(<LoadedView />).find(Row).find(Col);
      expect(columns.at(0).find(CourseList).length).toEqual(1);
    });
    it('displays WidgetSidebar in second column', () => {
      const columns = shallow(<LoadedView />).find(Row).find(Col);
      expect(columns.at(1).find(WidgetSidebar).length).toEqual(1);
    });
  };
  const testSnapshot = () => {
    test('snapshot', () => {
      expect(shallow(<LoadedView />)).toMatchSnapshot();
    });
  };
  describe('collapsed', () => {
    testColumns();
    testSnapshot();
    it('does not show spacer component above widget sidebar', () => {
      const columns = shallow(<LoadedView />).find(Col);
      expect(columns.at(1).find('h2').length).toEqual(0);
    });
  });

  describe('not collapsed', () => {
    beforeEach(() => { hooks.useIsDashboardCollapsed.mockReturnValueOnce(false); });
    testColumns();
    testSnapshot();
    it('shows a blank (nbsp) h2 spacer component above widget sidebar', () => {
      const columns = shallow(<LoadedView />).find(Col);
      // nonbreaking space equivalent
      expect(columns.at(1).find('h2').text()).toEqual('\xA0');
    });
  });
});
