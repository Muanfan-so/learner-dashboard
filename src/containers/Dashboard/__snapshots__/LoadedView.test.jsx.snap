// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadedView collapsed snapshot 1`] = `
<Container
  fluid={true}
  size="xl"
>
  <Row>
    <Col
      className="course-list-column"
      lg={
        Object {
          "offset": 0,
          "span": 12,
        }
      }
      md={
        Object {
          "offset": 1,
          "span": 10,
        }
      }
      sm={
        Object {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        Object {
          "offset": 0,
          "span": 8,
        }
      }
    >
      <CourseList />
    </Col>
    <Col
      className="sidebar-column"
      lg={
        Object {
          "offset": 0,
          "span": 12,
        }
      }
      md={
        Object {
          "offset": 1,
          "span": 10,
        }
      }
      sm={
        Object {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        Object {
          "offset": 0,
          "span": 4,
        }
      }
    >
      <WidgetSidebar />
    </Col>
  </Row>
</Container>
`;

exports[`LoadedView not collapsed snapshot 1`] = `
<Container
  fluid={true}
  size="xl"
>
  <Row>
    <Col
      className="course-list-column"
      lg={
        Object {
          "offset": 0,
          "span": 12,
        }
      }
      md={
        Object {
          "offset": 1,
          "span": 10,
        }
      }
      sm={
        Object {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        Object {
          "offset": 0,
          "span": 8,
        }
      }
    >
      <CourseList />
    </Col>
    <Col
      className="sidebar-column"
      lg={
        Object {
          "offset": 0,
          "span": 12,
        }
      }
      md={
        Object {
          "offset": 1,
          "span": 10,
        }
      }
      sm={
        Object {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        Object {
          "offset": 0,
          "span": 4,
        }
      }
    >
      <h2
        className="course-list-title"
      >
        Â 
      </h2>
      <WidgetSidebar />
    </Col>
  </Row>
</Container>
`;
